dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java

```diff
@@ -46,42 +53,75 @@
                 ForwardedIpAwareWebAuthenticationDetails.class.isAssignableFrom( details.getClass() ) )
             {
                 ForwardedIpAwareWebAuthenticationDetails authDetails = (ForwardedIpAwareWebAuthenticationDetails) details;
-                String ip = authDetails.getIp();
-
-                builder.append( "; ip: " );
-                builder.append( ip );
+                ipAddress = "; ip: " + authDetails.getIp();
+                sessionId = hashSessionId( authDetails.getSessionId() );
+            }
+            else if ( OAuth2LoginAuthenticationToken.class.isAssignableFrom( event.getAuthentication().getClass() ) )
+            {
+                OAuth2LoginAuthenticationToken authenticationToken = (OAuth2LoginAuthenticationToken) event
+                    .getAuthentication();
 
-                String sessionId = authDetails.getSessionId();
-                if ( sessionId != null )
+                DhisOidcUser principal = (DhisOidcUser) authenticationToken.getPrincipal();
+                if ( principal != null )
                 {
-                    HashCode hash = Hashing.sha256().newHasher().putString( sessionId, Charsets.UTF_8 ).hash();
-                    builder.append( " sessionId: " );
-                    builder.append( hash.toString() );
+                    UserCredentials userCredentials = principal.getUserCredentials();
+                    authName = userCredentials.getUsername();
                 }
 
+                WebAuthenticationDetails oauthDetails = (WebAuthenticationDetails) authenticationToken.getDetails();
+                ipAddress = "; ip: " + oauthDetails.getRemoteAddress();
+                sessionId = hashSessionId( oauthDetails.getSessionId() );
             }
-
-            if ( event instanceof AbstractAuthenticationFailureEvent )
+            else if ( OAuth2AuthenticationToken.class.isAssignableFrom( event.getSource().getClass() ) )
             {
-                builder.append( "; exception: " );
-                builder.append( ((AbstractAuthenticationFailureEvent) event).getException().getMessage() );
+                OAuth2AuthenticationToken authenticationToken = (OAuth2AuthenticationToken) event.getSource();
+                DhisOidcUser principal = (DhisOidcUser) authenticationToken.getPrincipal();
+
+                if ( principal != null )
+                {
+                    UserCredentials userCredentials = principal.getUserCredentials();
+                    authName = userCredentials.getUsername();
+                }
             }
 
-            log.warn( builder.toString() );
+            log.warn( eventClassName + "; username: " + authName + ipAddress + sessionId + exceptionMessage );
         }
     }
+
+    private String hashSessionId( String sessionId )
+    {
+        if ( sessionId == null )
+        {
+            return "";
+        }
+
+        HashCode hash = Hashing.sha256().newHasher().putString( sessionId, Charsets.UTF_8 ).hash();
```

Comment:
Does it make sense to cache this hasher? or even just use `Hashing.sha256().hashString(sessionId, StandardCharsets.UTF_8).toString()` direcly?

Reply:
I made the HashingFunction a field. The rest of the API is mostly immutable so object reuse is not possible.